runtime/lua/vim/treesitter/_query_linter.lua

```diff
@@ -0,0 +1,198 @@
+--- vim.tbl_contains can not be used because it does not do nested comparison (like vim.deep_equal)
+--- and the table returned by vim.treesitter.language.inspect is a mixed list/dictionary-like table
+local function table_contains(predicate, table)
+  for _, elt in pairs(table) do
+    if predicate(elt) then
+      return true
+    end
+  end
+  return false
+end
+
+---@private
+--- Determines the target language of a query file by its path: <lang>/<query_type>.scm
+local function guess_query_lang(buf, _)
+  local filename = vim.api.nvim_buf_get_name(buf)
+  if filename ~= '' then
+    local ok, query_lang = pcall(vim.fn.fnamemodify, filename, ':p:h:t')
+    return ok and query_lang
+  end
+end
+
+---@private
+local function normalize_opts(buf, opts)
+  opts = opts or {}
+  opts.langs = opts.langs or guess_query_lang
+  if type(opts.langs) == 'function' then
+    opts.langs = opts.langs(buf, opts)
+  end
+  if type(opts.langs) ~= 'table' then
+    opts.langs = { opts.langs }
+  end
+  return opts
+end
+
+local lint_query = [[;; query
+  (program [(named_node) (list) (grouping)] @toplevel)
+  (named_node
+    name: _ @node.named)
+  (anonymous_node
+    name: _ @node.anonymous)
+  (field_definition
+    name: (identifier) @field)
+  (predicate
+    name: (identifier) @predicate.name
+    type: (predicate_type) @predicate.type)
+  (ERROR) @error
+]]
+
+---@private
+function M.lint(buf, opts)
+  if buf == 0 then
+    buf = vim.api.nvim_get_current_buf()
+  end
+
+  local diagnostics = {}
+  local query = vim.treesitter.query.parse('query', lint_query)
+
+  opts = normalize_opts(buf, opts)
+
+  local predicate_list = vim.treesitter.query.list_predicates()
+  local negated_predicate_list = vim.tbl_map(function(p)
+    return 'not-' .. p
+  end, vim.treesitter.query.list_predicates())
+  local directive_list = vim.treesitter.query.list_directives()
+
+  for i, lang in ipairs(opts.langs or { nil }) do
+    local ok, parser_info = pcall(vim.treesitter.language.inspect, lang)
+    if not ok then
+      parser_info = nil
+    end
+
+    local parser = vim.treesitter.get_parser(buf)
+    parser:parse()
+    parser:for_each_tree(function(tree, ltree)
+      if ltree:lang() == 'query' then
+        for _, match, _ in query:iter_matches(tree:root(), buf, 0, -1) do
+          local predicate
+          local predicate_type
+          for id, node, _ in pairs(match) do
+            local cap_id = query.captures[id]
+
+            -- perform language-independent checks only for first lang
+            if cap_id == 'error' and i == 1 then
+              add_lint_for_node(diagnostics, node, buf, 'Syntax error')
+            end
+
+            -- other checks rely on Neovim parser introspection
+            if lang and parser_info then
+              if cap_id == 'toplevel' then
+                local query_text = vim.treesitter.get_node_text(node, buf)
+                if not parse_cache[lang] then
+                  parse_cache[lang] = {}
```

Comment:
The nesting is very deep here. Can this be broken down?

Reply:
better now?